[clang][SYCL] Add sycl_external attribute and restrict emitting device code

schittir opened the pull request:
This patch is part of the upstreaming effort for supporting SYCL language front end. 
It makes the following changes:
1. Adds sycl_external attribute for functions with external linkage, which is intended for use to implement the SYCL_EXTERNAL macro as specified by the SYCL 2020 specification
2. Adds checks to avoid emitting device code when sycl_external and sycl_kernel_entry_point attributes are not enabled
3. Fixes test failures caused by the above changes

This patch is missing diagnostics for the following diagnostics listed in the SYCL 2020 specification's section 5.10.1, which will be addressed in a subsequent PR: 
Functions that are declared using SYCL_EXTERNAL have the following additional restrictions beyond those imposed on other device functions:
1. If the SYCL backend does not support the generic address space then the function cannot use raw pointers as parameter or return types. Explicit pointer classes must be used instead;
2. The function cannot call group::parallel_for_work_item;
3. The function cannot be called from a parallel_for_work_group scope.

In addition to that, the subsequent PR will also implement diagnostics for inline functions including virtual functions defined as inline.

schittir committed:
Add sycl_external attribute

schittir committed:
Fix test and remove space

tahonermann commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12746,6 +12746,11 @@ def err_sycl_special_type_num_init_method : Error<
   "method defined">;
 
+//SYCL external attribute diagnostics
+def err_sycl_attribute_internal_decl
+    : Error<"%0 attribute cannot be applied to a %select{function|variable}1"
+            " without external linkage">;
```
The only other uses of this diagnostic in downstream Intel repositories are for Intel extensions, so I don't think we need a generalized diagnostic.

The SYCL 2020 specification doesn't support use of the `SYCL_EXTERNAL` macro with variables, so we don't need to support that (if we need such an extension, we might want to introduce a different attribute for it later).

The suggested change below includes renaming, removal of applicability to variables, rephrasing inspired from `err_attribute_selectany_non_extern_data`, and style changes to match other nearby declarations.
```suggestion
def err_sycl_external_invalid_linkage : Error<
  "'sycl_external' can only be applied to functions with external linkage">;
```

We'll eventually need to add diagnostics for at least the following scenarios as indicated in the SYCL 2020 specification in section [5.10.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:syclexternal). I haven't checked whether diagnostics are already present in downstream Intel report. For scenarios that are straight forward to diagnose, we should add diagnostics now. For others, we need to ensure we file an issue or somehow track adding diagnostics later. Ideally, we would add tests for each of these now with appropriate FIXME annotations as needed.
- The `sycl_external` attribute must be present on the first declaration of a function and may optionally be present on subsequent declarations.
- If the device target does not support the generic address space, then functions declared with `sycl_external` cannot have parameter types or a return type that involves a raw pointer.
- Functions declared with `sycl_external` cannot call `group::parallel_for_work_item()`.
- Functions declared with `sycl_external` cannot be called from a `parallel_for_work_group` scope.

schittir replied:
Planning to address this and add missing diagnostics in a subsequent PR.

tahonermann commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -408,10 +408,14 @@ class SubjectList<list<AttrSubject> subjects, SubjectDiag diag = WarnDiag,
   // The language option to test; ignored when custom code is supplied.
   string Name = name;
 
+  // If set to 1, the attribute is accepted but is silently ignored. This is
+  // useful in multi-compilation situations like SYCL.
+  bit SilentlyIgnore = silentlyIgnore;
+
```
None of the proposed changes query this bit (downstream Intel repos have additional code in `GenerateTestExpression()` in `clang/utils/TableGen/ClangAttrEmitter.cpp` that checks it).

I think we should hold off on adding this for now. Or, we should add it via a separate PR that justifies its use based on similar existing utility for CUDA, HIP, HLSL, etc...

tahonermann commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -1545,6 +1550,18 @@ def SYCLKernel : InheritableAttr {
+                                             "global variables">;
+
+def SYCLExternal : InheritableAttr {
+  let Spellings = [GNU<"sycl_external">];
```
Since this attribute is a Clang extension, I think we should limit it to the Clang spelling for now as is done for the `sycl_kernel_entry_point` attribute.
```suggestion
  let Spellings = [Clang<"sycl_external">];
```

tahonermann commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -1545,6 +1550,18 @@ def SYCLKernel : InheritableAttr {
+
+def SYCLExternal : InheritableAttr {
+  let Spellings = [GNU<"sycl_external">];
+  let Subjects = SubjectList<[Function, GlobalStorageNonLocalVar]>;
```
In SYCL 2020, the `SYCL_EXTERNAL` macro only applies to functions. Intel downstream repos allow the `sycl_device` attribute (which is used to implement `SYCL_EXTERNAL` there) to be applied to variables, but that isn't needed for SYCL 2020 conformance. If we later find a need to extend the semantics of `sycl_external` to variables, we should check with the SYCL WG to see if additions to the SYCL specification to support such semantics are warranted.

tahonermann commented on clang/test/SemaSYCL/sycl-external-attribute.cpp:
```diff
@@ -0,0 +1,52 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify -DSYCL %s
+// RUN: %clang_cc1 -fsycl-is-host -fsyntax-only -verify -DHOST %s
```
Since SYCL applies to both host and device, let's differentiate by "host" and "device" (with or without "SYCL")
```suggestion
// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify -DDEVICE %s
// RUN: %clang_cc1 -fsycl-is-host -fsyntax-only -verify -DHOST %s
```

tahonermann commented on clang/test/SemaSYCL/sycl-external-attribute.cpp:
```diff
@@ -0,0 +1,52 @@
+__attribute__((sycl_external)) // expected-warning {{'sycl_external' attribute ignored}}
+void baz() {}
+
+#endif
```
There are a few different concerns being exercised by this test. We have a precedent in place for exercising appertainment, grammar, and relevance by different test files in the `sycl-kernel-entry-point-attr-*` tests. Can we follow that pattern (and associated naming conventions) here? I'm not so concerned about the testing being as extensive as it is in those other tests, but splitting these concerns will allow additional cases to be added later with less disruption.

tahonermann commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -12926,6 +12929,10 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
 
     // FIXME: Functions declared with SYCL_EXTERNAL are required during
     // device compilation.
```
```suggestion
```

tahonermann reviewed (changes requested):
Thanks for working on this @schittir! I completed an initial pass of all of the code, but still need to look more closely at the documentation updates.

schittir commented:
> Thanks for working on this @schittir! I completed an initial pass of all of the code, but still need to look more closely at the documentation updates.

Thank you for the initial pass review, Tom!

schittir commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -12909,6 +12909,9 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
   if (D->hasAttr<WeakRefAttr>())
     return false;
 
+  if (LangOpts.SYCLIsDevice && !D->hasAttr<SYCLKernelEntryPointAttr>())
+    return false;
```
Yes, indeed. I left that one out of this draft because adding that check caused three additional failures than those caused by the check for KernelEntryPointAttr alone. I didn't expect you to review it, but I should have labeled it clearly, just in case. Thank you!

schittir committed:
Address review comments #1

schittir commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -12909,6 +12909,10 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
   if (D->hasAttr<WeakRefAttr>())
     return false;
 
+  if (LangOpts.SYCLIsDevice &&
+       (!D->hasAttr<SYCLKernelEntryPointAttr>() || !D->hasAttr<SYCLExternalAttr>()))
```
Good catch. Thank you!

schittir committed:
Fix conditional and failing tests

schittir committed:
Fix the remaining six failing tests

schittir committed:
Fix formatting

schittir committed:
Merge branch 'main' into sycl_external

schittir committed:
Remove sycl_external attribute support to variables.

Support for functions is sufficient for SYCL 2020 spec conformance.

schittir committed:
Rename test file

schittir committed:
Add tests for sycl_external attribute

schittir committed:
Add code examples to sycl_external documentation

Fznamznon commented on clang/test/CodeGenSYCL/amd-address-space-conversions.cpp:
```diff
@@ -1,128 +1,128 @@
 // RUN: %clang_cc1 -triple amdgcn-amd-amdhsa -fsycl-is-device -disable-llvm-passes -emit-llvm %s -o - | FileCheck %s
 void bar(int &Data) {}
-// CHECK: define dso_local void @[[RAW_REF:[a-zA-Z0-9_]+]](ptr noundef nonnull align 4 dereferenceable(4) %
+// CHECK-DAG: define {{.*}} void @[[RAW_REF:[a-zA-Z0-9_]+]](ptr noundef nonnull align 4 dereferenceable(4) %
```
We could actually check the calling convention and linkage instead of simply skipping it.

schittir replied:
I see this in place: https://github.com/intel/llvm/commit/1ea43e2a77667eab3d8702732afe62cf9f6347a9, and followed the pattern established here to help avoid downstream conflicts. 
Actually checking the calling convention and linkage makes sense too.

Fznamznon commented on clang/test/CodeGenSYCL/function-attrs.cpp:
```diff
@@ -29,21 +29,10 @@ int foo() {
 
 template <typename Name, typename Func>
-__attribute__((sycl_kernel)) void kernel_single_task(const Func &kernelFunc) {
+[[clang::sycl_external]] void kernel_single_task(const Func &kernelFunc) {
```
I think it was meant to be a kernel, so sycl_kernel_entry_point attribute would be appropriate here.
```suggestion
[[clang::sycl_kernel_entry_point]] void kernel_single_task(const Func &kernelFunc) {
```

schittir replied:
Agreed. It should have been [[clang::sycl_kernel_entry_point(Name)]]

Fznamznon commented on clang/test/SemaSYCL/sycl-external-attr-grammar.cpp:
```diff
@@ -0,0 +1,16 @@
+[[clang::sycl_external]] int b;
+};
+
+// FIXME: The first declaration of a function is required to have the attribute.
```
Why?

schittir replied:
SYCL spec 2020 section 5.10.0 describes it so: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:syclexternal

schittir committed:
Merge branch 'main' into sycl_external

schittir committed:
Update clang/lib/Sema/SemaDeclAttr.cpp

Co-authored-by: Mariya Podchishchaeva <mariya.podchishchaeva@intel.com>

schittir committed:
Address review comments -2

schittir commented on clang/test/CodeGenSYCL/unique_stable_name_windows_diff.cpp:
```diff
@@ -25,20 +40,29 @@ int main() {
   // CHECK: private unnamed_addr [[$ADDRSPACE]]constant [17 x i8] c"_ZTSZ4mainEUliE_\00"
   // CHECK: private unnamed_addr [[$ADDRSPACE]]constant [17 x i8] c"_ZTSZ4mainEUldE_\00"
+
+  // On Windows, ensure that we haven't broken the 'lambda numbering' for thex
```
Thank you for the review, @Fznamznon!

tahonermann reviewed (changes requested):
Thanks @schittir, here are my initial review comments. I expect to review the newly added tests more closely once these comments are all addressed.

schittir commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,38 @@
+[[clang::sycl_external]] int *func0() { return nullptr; }
+
+[[clang::sycl_external]] void func2(int *) {}
+
```
I created a tracker to account for the following missing diagnostics that need to be implemented to follow spec 2020 for this attribute. 

- When compiling for a SYCL device target that does not support the generic address space, the function shall not specify a raw pointer or reference type as the return type or as a parameter type.
- The function cannot call sycl::group::parallel_for_work_item.
- The function cannot be called from a parallel_for_work_group scope.

schittir replied:
I added these tests in the latest commit 

- main() should be rejected.
- Explicitly deleted functions should be rejected

schittir commented on clang/test/SemaSYCL/sycl-external-attr-ignore.cpp:
```diff
@@ -0,0 +1,12 @@
+// expected-warning@+1{{'sycl_external' attribute ignored}}
+[[clang::sycl_external]] void bar() {}
+
+// expected-warning@+1{{'sycl_external' attribute ignored}}
+[[clang::sycl_external]] int a;
```
This is for cases where the attribute is ignored when sycl is not enabled (i.e., using -fsycl-is-device option)  
It is a warning here too: clang/test/SemaSYCL/sycl-kernel-entry-point-attr-ignored.cpp 
Do you recommend changing these? 

I will change the file name to sycl-external-attr-ignored.cpp for consistency.

tahonermann replied:
Ah, I missed that this test was exercising non-SYCL compilation. Nevermind, this looks exactly right!

schittir committed:
Address review comments -3

schittir committed:
Rename test file

schittir committed:
Address review comments -4

schittir committed:
Merge branch 'main' into sycl_external

schittir committed:
Fix failing tests and address review comments

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -476,6 +476,41 @@ The SYCL kernel in the previous code sample meets these expectations.
+The attribute is intended for use in the implementation of the ``SYCL_EXTERNAL``
+macro as specified in section 5.10.1, "SYCL functions and member functions
+linkage", of the SYCL 2020 specification.
+The attribute only appertains to functions and only those that meet the
+following requirements.
+* Has external linkage.
+* Is not explicitly defined as deleted (the function may be an explicitly
+  defaulted function that is defined as deleted).
+The attribute shall be present on the first declaration of a function and
+may optionally be present on subsequent declarations.
+When compiling for a SYCL device target that does not support the generic
+address space, the function shall not specify a raw pointer or reference type
+as the return type or as a parameter type.
+See section 5.9, "Address-space deduction", of the SYCL 2020 specification.
+The following examples demonstrate the use of this attribute:
```
I think blank lines are required between paragraphs and lists.
```suggestion

The attribute only appertains to functions and only those that meet the
following requirements.

* Has external linkage.
* Is not explicitly defined as deleted (the function may be an explicitly
  defaulted function that is defined as deleted).

The attribute shall be present on the first declaration of a function and
may optionally be present on subsequent declarations.

When compiling for a SYCL device target that does not support the generic
address space, the function shall not specify a raw pointer or reference type
as the return type or as a parameter type.
See section 5.9, "Address-space deduction", of the SYCL 2020 specification.

The following examples demonstrate the use of this attribute:
```

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -476,6 +476,41 @@ The SYCL kernel in the previous code sample meets these expectations.
+
+.. code-block:: c++
+
+[[clang::sycl_external]] void Foo();
+
+[[clang::sycl_external]] void Bar() { /* ... */ }
+
+[[clang::sycl_external]] extern void Baz();
+
```
I'm not sure, but I think the code needs to be indented to match the ".." from the code-block. I think an additional example to illustrate the external linkage requirement would be useful too.
```suggestion
  [[clang::sycl_external]] void Foo(); // Ok.

  [[clang::sycl_external]] void Bar() { /* ... */ } // Ok.

  [[clang::sycl_external]] extern void Baz(); // Ok.

  [[clang::sycl_external]] static void Quux() { /* ... */ } // error:  Quux() has internal linkage.
```

tahonermann reviewed (changes requested):
Thanks for addressing most of my earlier comments. I took another quick look today and added a couple more comments.

schittir committed:
Address review comments -3

schittir committed:
Merge branch 'main' into sycl_external

schittir committed:
Merge branch 'main' into sycl_external

tahonermann commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -7206,6 +7206,9 @@ ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D, const ParsedAttr &AL,
     handleEnumExtensibilityAttr(S, D, AL);
     break;
+  case ParsedAttr::AT_SYCLExternal:
+    S.SYCL().handleExternalAttr(D, AL);
```
I think `handleSimpleAttribute()` can be called here and the diagnostics handled elsewhere. See additional comments in `SemaSYCL::handleExternalAttr()`.

tahonermann commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -202,6 +202,25 @@ void SemaSYCL::handleKernelAttr(Decl *D, const ParsedAttr &AL) {
+    Diag(AL.getLoc(), diag::err_sycl_attribute_avoid_main);
+    return;
+  }
+  if (FD->isDeleted()) {
+    Diag(AL.getLoc(), diag::err_sycl_attribute_avoid_deleted_function);
+    return;
+  }
```
I think this should call `isDeletedAsWritten()` so that it doesn't check for functions that are explicitly defaulted but defined as deleted. See additional test comments.

Also, I think this diagnostic would be better done in `Sema::CheckFunctionDeclaration()`; probably near the call to `CheckSYCLEntryPointFunctionDecl()`. Per my other comment, perhaps it makes sense to call a `CheckSYCLExternalFunctionDecl()` function there.

tahonermann commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -202,6 +202,25 @@ void SemaSYCL::handleKernelAttr(Decl *D, const ParsedAttr &AL) {
 }
 
+void SemaSYCL::handleExternalAttr(Decl *D, const ParsedAttr &AL) {
+  auto *FD = cast<FunctionDecl>(D);
+  if (!FD->isExternallyVisible()) {
+    Diag(AL.getLoc(), diag::err_sycl_attribute_invalid_linkage);
+    return;
+  }
```
I think the check for external linkage is better done in `Sema::CheckFunctionDeclaration()` so that other semantic checking of the function is done first. See the call to `CheckSYCLEntryPointFunctionDecl()` there; perhaps it makes sense to replicate that redirection to a `CheckSYCLExternalFunctionDecl()` function defined here.

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,54 @@
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func4(UnnX) {}
+
+// FIXME: The first declaration of a function is required to have the attribute.
+// The attribute may be optionally present on subsequent declarations
```
This diagnostic is pretty straight forward to implement. See `Sema::MergeFunctionDecl()` and the checks done for `CXX11NoReturnAttr` and `CarriesDependencyAttr`.

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,54 @@
+
+// FIXME: The first declaration of a function is required to have the attribute.
+// The attribute may be optionally present on subsequent declarations
+int foo(int c);
+
+[[clang::sycl_external]] void foo();
```
This is a little confusing. If the intent is specifically to exercise the attribute being present on one overload and not on another, then that is ok, but a comment would be helpful.

tahonermann reviewed (changes requested):
Additional comments for diagnostics and tests.

schittir committed:
Change diagnostic messages

schittir committed:
Merge branch 'main' into sycl_external

schittir committed:
Merge branch 'main' into sycl_external

schittir committed:
Change the second RUN line to use -sycl-is-host

schittir committed:
Switch to using sycl_external attr to pass the failing test

There seems to be an issue with the use of sycl_kernel_entry_point
attribute vs __builtin_unique_stable_name

schittir committed:
Revert RUN line to -fsycl-is-device

schittir committed:
Revert test change

schittir committed:
Merge branch 'main' into sycl_external

Fznamznon commented on clang/test/CodeGenSYCL/unique_stable_name_windows_diff.cpp:
```diff
@@ -25,20 +40,29 @@ int main() {
   // CHECK: private unnamed_addr [[$ADDRSPACE]]constant [17 x i8] c"_ZTSZ4mainEUliE_\00"
   // CHECK: private unnamed_addr [[$ADDRSPACE]]constant [17 x i8] c"_ZTSZ4mainEUldE_\00"
+
+  // On Windows, ensure that we haven't broken the 'lambda numbering' for the
+  // lambda itself.
+  // WIN: define internal void @"??R<lambda_1
+  // WIN: define internal void @"??R<lambda_2
+  // WIN: define internal void @"??R<lambda_3
+  // WIN: define internal void @"??R<lambda_4
```
Could you please revert all the changes made to this test?

schittir replied:
These changes were introduced to be consistent with changes downstream. I removed them now.

schittir committed:
Fix conflict resolution errors.

schittir committed:
Remove changes introduced from downstream.

schittir committed:
Update diagnostic messages in tests

schittir committed:
Undo more downstream changes

schittir committed:
Ungroup diagnostics and add test cases

schittir committed:
Merge branch 'main' into sycl_external

schittir committed:
Fix newly failing tests by adding sycl_external attribute

schittir committed:
Add constexpr and consteval test cases

schittir commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,54 @@
+[[clang::sycl_external]] int *func0() { return nullptr; }
+
+[[clang::sycl_external]] void func2(int *) {}
+
```
Added constexpr and consteval tests. 
Adding a warning for the consteval would make it more thorough, of course, but I don't know if there is value beyond that. I can certainly add it!

tahonermann replied:
Yeah, let's not bother with a warning for use with `consteval`. That would probably be considered excessive nannying.

tahonermann reviewed (changes requested):
This is looking good @schittir! I added some comments for minor issues. I still need to review the tests more closely.

schittir commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -4084,6 +4084,19 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
            diag::note_carries_dependency_missing_first_decl) << 0/*Function*/;
     }
 
+    // SYCL spec 2020
+    //   The first declaration of a function with external linkage must
+    //   specify sycl_external attribute.
+    //   Subsequent declarations may optionally specify this attribute.
```
Thank you for the suggestion. 
I will implement the wording as it is except for one point - wouldn't it be better to use `sycl_external` attribute instead of `SYCL_EXTERNAL` macro which is not yet implemented?

tahonermann replied:
I tend towards quoting the text verbatim. Readers are likely to be aware that the intent of the `sycl_external` attribute is to implement the `SYCL_EXTERNAL` macro; that was made clear in the documentation for the attribute.

schittir committed:
Use existing diagnostic and address other minor comments

tahonermann reviewed (changes requested):
A few more comments. I still haven't finished my review of `sycl-external-attr.cpp`, but will try to later today or over the weekend.

schittir commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12879,6 +12879,15 @@ def err_sycl_special_type_num_init_method : Error<
+// SYCL external attribute diagnostics
+def err_sycl_attribute_invalid_linkage : Error<
+  "'sycl_external' can only be applied to functions with external linkage">;
+def err_sycl_attribute_avoid_main : Error<
+  "'sycl_external' cannot be applied to the 'main' function">;
+def err_sycl_attribute_avoid_deleted_function
+    : Error<"'sycl_external' cannot be applied to an explicitly deleted "
+            "function">;
```
The wrapping and indentation style was modified by git-clang-format tool, which surprised me! I'll revert it.

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,70 @@
+
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func4(UnnX) {}
+
```
Other interesting internal linkage cases:

The attribute from the primary template is inherited by the specialization instantiated by the explicit template instantiation declaration. It could be argued that, because the explicit instantiation declaration names the symbol that causes the instantiated specialization to have internal linkage, a diagnostic should be produced here. However, other template shenanigans can produce similar results and would be difficult to diagnose well. I think we should ensure that a diagnostic is not issued for this case.
```
namespace { struct S6 {}; }
template<typename>
[[clang::sycl_external]] void func6() {}
template void func6<S6>();
```

In this case, the attribute from the primary template is not inherited by the specialization produced by the explicit specialization declaration. No diagnostic should be issued for this case. However, Clang doesn't implement the attribute inheritance semantics correctly (according to the standard), so if a diagnostic is issued, we should just add a FIXME comment and not worry about it.
```
namespace { struct S7 {}; }
template<typename>
[[clang::sycl_external]] void func7();
template<> void func7<S7>() {}
```

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,70 @@
+[[clang::sycl_external]] // expected-error {{'sycl_external' can only be applied to functions with external linkage}}
+  void func4(UnnX) {}
+
+// The first declaration of a SYCL external function is required to have this attribute.
+int foo(); // expected-note {{previous declaration is here}}
+
+[[clang::sycl_external]] int foo(); // expected-error {{'clang::sycl_external' attribute does not appear on the first declaration}}
```
Avoid blank lines between declarations that are part of the same test so that it is easy to see that they go together. Use line markers and references in expected diagnostics to avoid line length issues.
```suggestion
// The first declaration of a SYCL external function is required to have this attribute.
// expected-error@+3 {{'clang::sycl_external' attribute does not appear on the first declaration}}
// expected-note@#func8-decl {{previous declaration is here}}
int func8(); // #func8-decl
[[clang::sycl_external]] int func8();
```

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,70 @@
+  return 0;
+}
+
+class D {
+  [[clang::sycl_external]] void del() = delete; // expected-error {{'sycl_external' cannot be applied to an explicitly deleted function}}
+};
```
This test doesn't require a member function, but the presence of class `D` suggests that there is something specific to member functions here.
```suggestion
// expected-error {{'sycl_external' cannot be applied to an explicitly deleted function}}
[[clang::sycl_external]] void del() = delete;
```

tahonermann replied:
For the `sycl_kernel_entry_point` attribute, I had to add some checks, including the one for a deleted function, directly to `Sema::ActOnFinishFunctionBody`. See https://github.com/llvm/llvm-project/blob/711132dfa43211bce312db01a473aed51bae7b8c/clang/lib/Sema/SemaDecl.cpp#L16246-L16276. In my case, I had to check `isDeleted()`, but your existing code is correct in checking for `isDeletedAsWritten()`.

schittir replied:
Thank you! That fixes this issue.

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,70 @@
+  [[clang::sycl_external]]
+  A() {}
+
+  [[clang::sycl_external]] void func3() {}
```
Validate a static member function too.
```suggestion
  [[clang::sycl_external]] void mf();
  [[clang::sycl_external]] static void smf();
```

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,70 @@
+  [[clang::sycl_external]] virtual void bar() = 0;
+};
+
+[[clang::sycl_external]] int *func0() { return nullptr; }
+
+[[clang::sycl_external]] void func2(int *) {}
```
I assume these are intended to exercise the generic address space limitations. Let's add a comment to make that clear. I don't know if any of the target devices we expect to support lack support for the generic address space, so we might not be able to effectively test a positive diagnostic, but it is good to ensure no diagnostic for the common case.
```suggestion
// Devices that do not support the generic address space shall not specify
// a raw pointer or reference type as the return type or as a parameter type.
[[clang::sycl_external]] int *func0();
[[clang::sycl_external]] int &func1();
[[clang::sycl_external]] int &&func2();
[[clang::sycl_external]] void func3(int *);
[[clang::sycl_external]] void func4(int &);
[[clang::sycl_external]] void func5(int &&);
template<typename T>
[[clang::sycl_external]] void func6(T) {}
template void func6(int *);
template<> [[clang::sycl_external]] void func6<int*>(int *) {}
```

schittir replied:
`template<> [[clang::sycl_external]] void func6<long*>(long *) {}`

tahonermann reviewed (changes requested):
I finished an initial pass through all of the tests. I'll do another round once you've had a chance to catch up on the comments so far.

schittir commented on clang/test/SemaSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,70 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+// RUN: %clang_cc1 -fsycl-is-device -std=c++20 -fsyntax-only -verify -DCPP20 %s
```
I am assuming that you also intended to have the code guarded with CPP17 in the body.

schittir committed:
Add additional test cases and address review comments

tahonermann commented on clang/test/SemaSYCL/sycl-external-attr-grammar.cpp:
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+
+// expected-error@+1{{'clang::sycl_external' attribute only applies to functions}}
+[[clang::sycl_external]] int a;
+
+
+// expected-error@+2{{'clang::sycl_external' attribute only applies to functions}}
+struct s {
+[[clang::sycl_external]] int b;
+};
+
+// expected-error@+1{{'clang::sycl_external' attribute takes no arguments}}
+[[clang::sycl_external(3)]] void bar() {}
+
+// FIXME: The first declaration of a function is required to have the attribute.
+// The attribute may be optionally present on subsequent declarations
+int foo(int c);
+
+[[clang::sycl_external]] void foo();
```
I think those are useful to ensure that a diagnostic of some kind is issued. Ideally, I think the diagnostic would be something along the lines of "does not take an argument" since there is no defined argument syntax. I don't think it is worth trying to do better than the "expected expression" error though. The fact that some code some where is trying to parse arguments here is a separate and out-of-scope issue as far as I'm concerned.

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -476,6 +476,48 @@ The SYCL kernel in the previous code sample meets these expectations.
+When compiling for a SYCL device target that does not support the generic
+address space, the function shall not specify a raw pointer or reference type
+as the return type or as a parameter type.
+See section 5.9, "Address-space deduction", of the SYCL 2020 specification.
```
Oops, it looks like I quoted the wrong section of the specification previously. This should be:
```suggestion
See section 5.10, "SYCL offline linking", of the SYCL 2020 specification.
```

schittir committed:
Add test cases

tahonermann reviewed (changes requested):
This is looking great @schittir! I just realized that, since we made the change to allow the attribute to be present during host compilation that we should no longer restrict diagnostics to device compilation and that tests should exercise host compilation as well.

I suggested a few FIXME comments, but other than that, I think this is good to go!

schittir committed:
Add FIXME comments, enable diagnostics for host, remove a needless decl

schittir committed:
Merge branch 'main' into sycl_external

tahonermann reviewed (approved):
I love it @schittir, this looks great!

@erichkeane, @bader, please review and lend your approval. Note that some of the changes to the tests are to synchronize them with continued evolution that happened in DPC++ but was not previously upstreamed.

bader reviewed (commented):
LGTM in general. I have a few minor comments related to the code style and one related to CodeGen testing.

tahonermann commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -249,6 +249,19 @@ static bool CheckSYCLKernelName(Sema &S, SourceLocation Loc,
   return false;
 }
+void SemaSYCL::CheckSYCLExternalFunctionDecl(FunctionDecl *FD) {
+  for (auto *SEAttr : FD->specific_attrs<SYCLExternalAttr>()) {
```
I just happened to notice this while scrolling for Alexey's comments. It isn't necessary to use a `for` loop here since there is no meaningful distinction between one `sycl_external` attribute and another; unlike `sycl_kernel_entry_point`, there is no attribute argument that can cause them to be different. We should be able to just get the attribute.
```suggestion
  const auto *SEAttr = FD->getAttr<SYCLExternalAttr>();
  assert(SEPAttr && "Missing sycl_external attribute");
```

tahonermann commented on clang/test/CodeGenSYCL/unique_stable_name.cpp:
```diff
@@ -1,22 +1,22 @@
-// CHECK: @[[LAMBDA_NO_DEP:[^\w]+]] = private unnamed_addr addrspace(1) constant [[NO_DEP_LAMBDA_SIZE:\[[0-9]+ x i8\]]] c"_ZTSZ13lambda_no_depIidEvT_T0_EUlidE_\00",
-// CHECK: @[[LAMBDA_TWO_DEP:[^\w]+]] = private unnamed_addr addrspace(1) constant [[DEP_LAMBDA1_SIZE:\[[0-9]+ x i8\]]] c"_ZTSZ14lambda_two_depIZZ4mainENKUlvE0_clEvEUliE_ZZ4mainENKS0_clEvEUldE_EvvEUlvE_\00",
-// CHECK: @[[LAMBDA_TWO_DEP2:[^\w]+]] = private unnamed_addr addrspace(1) constant [[DEP_LAMBDA2_SIZE:\[[0-9]+ x i8\]]] c"_ZTSZ14lambda_two_depIZZ4mainENKUlvE0_clEvEUldE_ZZ4mainENKS0_clEvEUliE_EvvEUlvE_\00",
+// RUN: %clang_cc1 -triple x86_64-linux-pc -fsycl-is-host -disable-llvm-passes -emit-llvm %s -o - | FileCheck %s
```
There are essentially two changes to this test:

1. Incorporate changes to this test in intel/llvm that were never upstreamed. Compare with https://github.com/intel/llvm/blob/7756d71f2d3d617e583de1742de1f96cd7c67a50/clang/test/CodeGenSYCL/unique_stable_name.cpp.
2. Switch the test to use `sycl_kernel_entry_point` rather than `sycl_kernel` because the latter effectively has no implementation. Note that, prior to this PR, all functions emitted for host compilation were also emitted for device compilation. This test previously passed with `-fsycl-is-device` because functions were (incorrectly) being emitted. This test is intended to exercise `__builtin_sycl_unique_stable_name()` which is intended for use during host compilation.

Many of the `CodeGenSYCL` tests this PR changes add implicit validation of the `sycl_external` attribute since the additions of it were needed to keep these tests passing. `clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp` validates that `main` is not emitted during device compilation, but it does look like we are missing a dedicated test to validate that symbols that shouldn't be emitted during device compilation aren't unless declared with `sycl_external`. I think we should add such a test.

schittir replied:
Added a case in `clang/test/CodeGenSYCL/sycl-external-attr.cpp` where symbols are not emitted without sycl_external attribute

erichkeane reviewed (commented):
Ah shoot, apparently I forgot to click 'submit' when I reviewed this last week.  Sorry about that.  Here are my pending comments.

tahonermann commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -1641,6 +1641,13 @@ def DeviceKernel : DeclOrTypeAttr {
 }
 
+def SYCLExternal : InheritableAttr {
+  let Spellings = [Clang<"sycl_external">];
```
We use the `Clang` spelling for `sycl_kernel_entry_point`, what would be the reason for doing differently here? The attribute is an implementation detail used to provide the `SYCL_EXTERNAL` functionality, so shouldn't be directly written by SYCL programmers. Are you suggesting excluding the `GNU` spelling?

SYCL is only relevant for C++, so the `C23` spelling wouldn't be desired.

If we switch to (only) the `CXX11` spelling, I think the `clang` namespace should be retained.
```suggestion
  let Spellings = [CXX11<"clang", "sycl_external">];
```

tahonermann replied:
Ok, I'm fine with dropping the GNU spelling. But in that case, we should also change the `sycl_kernel_entry_point` attribute to match (either in this PR or in a separate one).

schittir replied:
Separate PR it is! 
I'll follow this spelling: `let Spellings = [CXX11<"clang", "sycl_external">];`

tahonermann commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12908,6 +12908,14 @@ def err_sycl_special_type_num_init_method : Error<
 
+// SYCL external attribute diagnostics
+def err_sycl_attribute_invalid_linkage : Error<
+  "'sycl_external' can only be applied to functions with external linkage">;
```
Searching for "can only" in `DiagnosticSemaKinds.td` doesn't suggest a consistent formulation. There are quite a few examples that match the currently proposed spelling.

If "attribute" is added after `sycl_external`, then I would suggest adding "the" before it; "the 'sycl_external' attribute ...". 

We probably should use `'clang::sycl_external'` for consistency with diagnostics that use substitution to provide the attribute name; see the "... attribute ignored" diagnostics for example.

schittir replied:
I'll use `let Spellings = [CXX11<"clang", "sycl_external">]; ` and separate PR it is!

tahonermann commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -4084,6 +4084,19 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
+    //   Redeclarations of the function in the same translation unit may
+    //   optionally use SYCL_EXTERNAL, but this is not required.
+    const SYCLExternalAttr *SEA = New->getAttr<SYCLExternalAttr>();
+    if (SEA && !Old->hasAttr<SYCLExternalAttr>()) {
```
That might be reasonable. I think we might still have to make it an error if there was an ODR-use before the definition though. I'd have to think about that more.

schittir replied:
@tahonermann - do you have any additional thoughts here? 
Thank you!

tahonermann replied:
Resolving this comment; warnings are now issued instead of errors for the attribute not appearing on the first declaration.

tahonermann commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -16275,6 +16297,13 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
 
+  if (FD && !FD->isInvalidDecl() && FD->hasAttr<SYCLExternalAttr>()) {
+    SYCLExternalAttr *SEAttr = FD->getAttr<SYCLExternalAttr>();
+    if (FD->isDeletedAsWritten())
```
Sindhu and I previously discussed that and concluded that would be undesirable in template scenarios. In non-dependent cases, diagnosing a defaulted-as-deleted case would probably be ok, but seemed enough of an edge case not to bother since an ODR-use of the deleted function would be diagnosed anyway.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -12937,6 +12937,10 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
     return false;
 
+  if (LangOpts.SYCLIsDevice && !D->hasAttr<SYCLKernelEntryPointAttr>() &&
+      !D->hasAttr<SYCLExternalAttr>())
```
Ah yes, part of why I find this solution problematic to read!  It ends up being much less obvious WHAT is missing.

erichkeane replied:
An additional thought after clicking 'start-review':  Perhaps something like:

```
// Comment explaining that on a sycl device, the ONLY thing we emit is functions (not variables), and ONLY those that have one of these two attributes.
if (LangOpts.SYCLIsDevice)
return isa<FunctionDecl>(D) && (D->hasAttr<...>() || D->hasAttr<....>());
```

I'm realizing as I look through this more, part of my problem is the vascus relationship between this section, the fact that these two attributes are only legal on a function decl, and the part on line 12948.  So there are three sizable parts of logic scattered around the program that makes this REALLY awful to figure out.

tahonermann replied:
Ok, that seems quite reasonable, I'm happy with that suggestion.

And yes, this sequence of if statements that are maybe-order-dependent-or-maybe-not makes for a pretty [fragile](https://www.youtube.com/watch?v=lR29zxrY2LM) house of cards. I'm sure there is a better way (maybe multiple better ways) to handle all of this, but it isn't obvious to me what better would look like. Either way, something for some future PR.

schittir committed:
Add test cases and address comments

schittir commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -12251,6 +12264,9 @@ bool Sema::CheckFunctionDeclaration(Scope *S, FunctionDecl *NewFD,
     SYCL().CheckSYCLEntryPointFunctionDecl(NewFD);
 
+  if (NewFD->hasAttr<SYCLExternalAttr>())
+    SYCL().CheckSYCLExternalFunctionDecl(NewFD);
```
```
 namespace { struct S9 {}; }
 struct T9 {
   using type = S9;
 };
 template<typename>
 [[clang::sycl_external]] void func9() {} // error here : {{'sycl_external' can only be applied to functions with external linkage}}
 template<typename T>
 [[clang::sycl_external]] void test_func9() {
   func9<typename T::type>(); // note here: {{in instantiation of function template specialization 'func9<(anonymous namespace)::S9>' requested here}}
 }
 template void test_func9<T9>(); // note here: {{in instantiation of function template specialization 'test_func9<T9>' requested here}}
 ```

This case is being diagnosed at present. Will fix it.

tahonermann replied:
To test for implicit instantiation, call `FunctionDecl::getTemplateSpecializationInfo()`. If it returns non-null, call `getSpecializationKind()` via the returned pointer. If that returns `TSK_ImplicitInstantiation`, then skip the diagnostic.

tahonermann replied:
@schittir and I discussed this and looked at some code together. Her checks for `TSK_ImplicitInstantiation` look right to me, but don't seem to be excluding explicit specializations and explicit instantiations for some reason. I'm fine with addressing this in a follow up PR if everyone else is. We'll likely need to do some debugging to really figure out what is going wrong.

tahonermann replied:
I debugged a bit. It looks like `getSpecializationKind()` is always returning `TSK_ImplicitInstantiation` at the point where we are checking it. I added a dump of the function declaration at the time of diagnosis and compared that to what is generated by `-Xclang -ast-dump`. Note the same addresses below, but the different specialization information.

Dumped decl:
```
FunctionDecl 0xdf89eb0 <.../sycl-external-attr.cpp:25:26, col:40> col:31 func6 'void ()' implicit_instantiation
|-TemplateArgument type '(anonymous namespace)::S6'
| `-RecordType 0xdf67730 '(anonymous namespace)::S6'
|   `-CXXRecord 0xdf67698 'S6'
`-SYCLExternalAttr 0xdf89fa8 <col:3, col:10>
```
Output from `-Xclang -ast-dump`:
```
| `-FunctionDecl 0xdf89eb0 <col:26, col:40> col:31 func6 'void ()' explicit_instantiation_definition
|   |-TemplateArgument type '(anonymous namespace)::S6'
|   | `-RecordType 0xdf67730 '(anonymous namespace)::S6'
|   |   `-CXXRecord 0xdf67698 'S6'
|   |-CompoundStmt 0xdf67b10 <col:39, col:40>
|   `-SYCLExternalAttr 0xdf89fa8 <col:3, col:10>
```

tahonermann replied:
Yes, that is exactly what is happening.

For an explicit instantiation, the template specialization kind isn't set (via `FunctionDecl::setTemplateSpecializationKind()` until near the end of `Sema::ActOnExplicitInstantiation()`. That seems somewhat unfortunate but there appears to be some interactions with the `dllimport` attribute. Certainly not something we should address in this PR.

For an explicit specialization, the template specialization kind is set in time, but we have the long standing issue that a declaration is implicitly instantiated and becomes the previous declaration of the explicit specialization which inherits attributes. We're clearly not going to address that in this PR.

Thinking about this some more, I don't think we should diagnose explicit template instantiations (either declarations or definitions). Note that explicit template instantiation doesn't allow for attributes to be specified; these are not syntactically valid:
```
extern template [[clang::sycl_external]] void ft<some-type>();
template [[clang::sycl_external]] void ft<some-type>();
```

I think we have a couple more tests to add (I'll add separate comments for those), but I think this is going to work out how we want it to (and if we don't diagnose the explicit template specializations, it isn't a big deal anyway).

schittir replied:
Thank you! I have all of this noted for a follow-up PR.

schittir committed:
Add test case where device code is not emitted without the attribute

schittir committed:
Merge branch 'main' into sycl_external

tahonermann reviewed (changes requested):
Thanks for the updates, @schittir. A few minor requests.

schittir commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,73 @@
+[[clang::sycl_external]] void tFunc1<int>(int arg) {}
+// CHECK: define dso_local spir_func void @_Z6tFunc1IiEvT_
+
+// FIXME: symbols should be emitted for the instantiation and the specialization
+// tFunc2 below
+template <typename T>
+void tFunc2(T arg) {}
+template void tFunc2<int>(int arg); // emit code for this
+template<> void tFunc2<char>(char arg) {} // and this
```
My bad. This case works as expected.

schittir committed:
Addressing further comments

schittir commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,73 @@
+};
+
+class B {
+  [[clang::sycl_external]] virtual void BFunc1WithAttr() { int i = 1; }
+// CHECK: define linkonce_odr spir_func void @_ZN1B14BFunc1WithAttrEv
```
Thank you for this comment, Tom. 
Per our discussion, I am noting this to address in a subsequent PR since this requires addressing filtering for inline and virtual functions elsewhere, potentially in ASTContext.cpp.

schittir committed:
Demote missing on first decl error to warning

schittir committed:
Fix tests that failed due to spell change to CXX11

schittir committed:
Add warning group

tahonermann commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -4092,6 +4092,19 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
+      Diag(SEA->getLocation(), diag::warn_attribute_missing_on_first_decl)
+          << SEA;
+      Diag(Old->getLocation(), diag::note_previous_declaration);
+      New->dropAttr<SYCLExternalAttr>();
```
Per my other comment with the diagnostic definition, I don't think we should ignore the attribute if it is not present on the first declaration.
```suggestion
```

tahonermann commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12910,6 +12910,18 @@ def err_sycl_special_type_num_init_method : Error<
+  "'clang::sycl_external' cannot be applied to the 'main' function">;
+def err_sycl_attribute_invalid_deleted_function : Error<
+  "'clang::sycl_external' cannot be applied to an explicitly deleted function">;
+def warn_attribute_missing_on_first_decl : Warning<
+  "%0 attribute does not appear on the first declaration">,
+  InGroup<IgnoredAttributes>;
```
I don't think we should ignore the attribute if it is not present on the first declaration since we can still honor its semantic effects. Assuming agreement, this warning shouldn't be a member of the `IgnoredAttributes` group.

Additionally, this diagnostic has either an odd name or an odd placement since it is colocated with `sycl_external` diagnostics, but is not named to be specific to them. I recommend giving it a `sycl_external` specific name to make it easier to search for.
```suggestion
def warn_sycl_external_missing_on_first_decl : Warning<
  "%0 attribute does not appear on the first declaration">;
```

schittir replied:
Did you mean to suggest removing a grouping altogether? 
It is my understanding that every warning without a -W flag must belong to a group, else, it needs to be added to clang/test/Misc/warning-flags.c; the test fails otherwise.

tahonermann commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12910,6 +12910,18 @@ def err_sycl_special_type_num_init_method : Error<
   "method defined">;
 
+// SYCL external attribute diagnostics
+def err_sycl_attribute_invalid_linkage : Error<
+  "'clang::sycl_external' can only be applied to functions with external"
+  " linkage">;
+def err_sycl_attribute_invalid_main : Error<
+  "'clang::sycl_external' cannot be applied to the 'main' function">;
+def err_sycl_attribute_invalid_deleted_function : Error<
+  "'clang::sycl_external' cannot be applied to an explicitly deleted function">;
```
For consistency with changes made in https://github.com/llvm/llvm-project/pull/151405 for the `sycl_kernel_entry_point` attribute, let's use `%0` to name the attribute. I also suggest we rename these diagnostics to be specific to `sycl_external`; we'll likely have other SYCL attributes in the future.
```suggestion
def err_sycl_external_invalid_linkage : Error<
  "%0 can only be applied to functions with external linkage">;
def err_sycl_external_invalid_main : Error<
  "%0 cannot be applied to the 'main' function">;
def err_sycl_external_invalid_deleted_function : Error<
  "%0 cannot be applied to an explicitly deleted function">;
```

tahonermann commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -250,6 +250,20 @@ static bool CheckSYCLKernelName(Sema &S, SourceLocation Loc,
+void SemaSYCL::CheckSYCLExternalFunctionDecl(FunctionDecl *FD) {
+  const auto *SEAttr = FD->getAttr<SYCLExternalAttr>();
+  assert(SEAttr && "Missing sycl_external attribute");
+  if (!FD->isExternallyVisible()) {
+    Diag(SEAttr->getLocation(), diag::err_sycl_attribute_invalid_linkage);
+    return;
+  }
```
An early return should not be done here since that would prevent other possible diagnostics.
```suggestion
  if (!FD->isExternallyVisible())
    Diag(SEAttr->getLocation(), diag::err_sycl_attribute_invalid_linkage);
```

tahonermann commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -250,6 +250,20 @@ static bool CheckSYCLKernelName(Sema &S, SourceLocation Loc,
+    Diag(SEAttr->getLocation(), diag::err_sycl_attribute_invalid_linkage);
+    return;
+  }
+  if (FD->isDeletedAsWritten()) {
+    Diag(SEAttr->getLocation(),
+         diag::err_sycl_attribute_invalid_deleted_function);
+    return;
+  }
```
An early return should not be done here since that would prevent other possible diagnostics (that might be added later in this function in the future).
```suggestion
  if (FD->isDeletedAsWritten())
    Diag(SEAttr->getLocation(),
         diag::err_sycl_attribute_invalid_deleted_function);
```

tahonermann commented on clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp:
```diff
@@ -100,11 +100,8 @@ int main() {
-// CHECK-DEVICE:      Function Attrs: convergent mustprogress noinline norecurse nounwind optnone
-// CHECK-DEVICE-NEXT: define {{[a-z_ ]*}}noundef i32 @main() #0
+// main() shouldn't be emitted in device code.
+// CHECK-NOT: define {{[a-z_ ]*}}noundef i32 @main() #0
```
Can this be simplified to be less sensitive to the presence of keywords? Can we just look for `@main()`?
```suggestion
// CHECK-NOT: @main()
```

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+
+// This test code generation when sycl_external attribute is used
+
+// Function defined and not used - symbols emitted
```
```suggestion
// Function defined and not used - definition emitted
```

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+[[clang::sycl_external]] int square(int x) { return x*x; }
+// CHECK: define dso_local spir_func noundef i32 @_Z6squarei
+
+// Function defined and used - symbols emitted
```
```suggestion
// Function defined and used - definition emitted
```

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+[[clang::sycl_external]] int squareUsed(int x) { return x*x; }
+// CHECK: define dso_local spir_func noundef i32 @_Z10squareUsedi
+
+// Constexpr function defined and not used - symbols emitted
```
Let's note this as a FIXME since there is no need to emit an unused inline function.
```suggestion
// FIXME: Constexpr function defined and not used - no definition emitted
```

schittir replied:
I was under the opposite impression per one of the older review comments but I can't find it at the moment. No related wording in the spec either.

tahonermann replied:
Functions declared `constexpr` are implicitly `inline` and are therefore available with the same definition in every translation unit that declares them (under penalty of the ODR). We therefore shouldn't have to emit an unused one.

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+// CHECK-NOT: define {{.*}} i32 @_Z8declOnlyv
+
+
+// FIXME: Function declared and used but not defined - emit external reference
```
This FIXME is surprising. If the function is used, a declaration must be present to satisfy the use.
```suggestion
// FIXME: Function declared and used but not defined - emit external reference
```

Ah, the issue is that the function is only used on the host, not on the device. We should test both of those scenarios (with different functions declared with the attribute).

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+// FIXME: Function declared and used but not defined - emit external reference
+[[clang::sycl_external]] void declUsed(int y);
+
+// Function declared with the attribute and later defined - symbols emitted
```
```suggestion
// Function declared with the attribute and later defined - definition emitted
```

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+// CHECK: define dso_local spir_func noundef i32 @_Z5func1i
+
+class A {
+// Defaulted special member functions - no symbols emitted
```
```suggestion
// Unused defaulted special member function - no symbols emitted
```

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+};
+
+class B {
+  [[clang::sycl_external]] virtual void BFunc1WithAttr() { int i = 1; }
```
Since `virtual` functions are not allowed in device code in SYCL 2020, I think we should diagnose uses of `sycl_external` with them.

schittir replied:
Ah! Good catch. Thank you.

schittir replied:
Undoing `[the change](https://github.com/llvm/llvm-project/pull/140282/commits/d1906b5c99c4a98c3a1314571898fc2fd2b689fd)` to diagnose virtual functions per conversation with @tahonermann in light of https://github.com/KhronosGroup/SYCL-Docs/issues/565; this case is currently supported downstream.

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+};
+
+class C {
+// Special member function defined - symbols emitted
```
```suggestion
// Special member function defined - definition emitted
```

tahonermann commented on clang/test/CodeGenSYCL/sycl-external-attr.cpp:
```diff
@@ -0,0 +1,77 @@
+// CHECK: define linkonce_odr spir_func void @_ZN1CD1Ev
+};
+
+int ret1() { return 1; }
+[[clang::sycl_external]] int withAttr() { return ret1(); }
+// CHECK: define dso_local spir_func noundef i32 @_Z8withAttrv
+// CHECK: define dso_local spir_func noundef i32 @_Z4ret1v
```
```suggestion
// Function reachable from an unused function - definition emitted
int ret1() { return 1; }
[[clang::sycl_external]] int withAttr() { return ret1(); }
// CHECK: define dso_local spir_func noundef i32 @_Z8withAttrv
// CHECK: define dso_local spir_func noundef i32 @_Z4ret1v
```

tahonermann reviewed (changes requested):
Another batch of rather small nit-picky things.

Things left to follow up on appear to include:
- Fixing diagnostics sensitive to implicit instantiation, explicit instantiation, and explicit specialization; might be done in a follow up PR.
- Improved handling of whether inline functions are emitted; might be done in a follow up PR.

schittir committed:
Address comments; added diagnostic for virtual function

schittir committed:
Provide attribute as argument for diagnostics

erichkeane commented:
I'm a bit behind on this, but am happy when Tom is.  Please make sure these followups happen ASAP/Next, they are all critical, but this review/conversation is getting too long to have them happen here.

schittir committed:
Undo diagnostic for virtual function

schittir committed:
Add "nonportable-sycl" diagnostic group definition

schittir committed:
Remove diagnostic for implicit template instantiation;

Address other review comments

tahonermann reviewed (changes requested):
Suggestions to address the diagnostic issues with explicit and implicit template instantiations as well as a couple of additional tests.

tahonermann reviewed (changes requested):
Thanks, @schittir. I promise these are my last requests. I went back through all the resolved conversations and I am quite content to approve this after these two comments are addressed.

schittir committed:
Undo accidental delete and add a virtual function case

tahonermann reviewed (approved):
Thank you, @schittir, for all your hard work on this! It looks great!

@erichkeane, @bader, please review at your earliest convenience!

schittir commented:
Hi @erichkeane, could you please take a look at this when you get a chance? Thank you!

schittir commented:
> > Hi @erichkeane, could you please take a look at this when you get a chance? Thank you!
> 
> See:
> 
> > I'm a bit behind on this, but am happy when Tom is. Please make sure these followups happen ASAP/Next, they are all critical, but this review/conversation is getting too long to have them happen here.

Thank you!
